### PR TITLE
Implement `requirement.json` for injection

### PR DESCRIPTION
**Pull request by TonyCTHsu**
**Change log entry**

None

**What does this PR do?**

Add `requirement.json` for injection.

**Motivation:**

Prevent injection when:

- we know system dependencies (libc, arch) aren't satisfied
- we know there's no point in injecting into a program (not useful, not effective, blocked by security policy)
- we know injecting into a program would be detrimental (breaking it due to incompatibility or security policy)

**Additional Notes:**

**JIRA**

- [INPLAT-103](https://datadoghq.atlassian.net/browse/INPLAT-103)

**How to test the change?**

Should be tested by CI in the added JSON test cases.

[INPLAT-103]: https://datadoghq.atlassian.net/browse/INPLAT-103?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib-injection/requirements.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
+  "$schema": "https://raw.githubusercontent.com/DataDog/auto_inject/refs/heads/main/preload_go/cmd/library_requirements_tester/testdata/requirements_schema.json",
+  "version": 1,
+  "native_deps": {
+    "glibc": [
+      {
+        "arch": "x86",
+        "supported": true,
+        "min": "2.28"
```
Comment (raphaelgavache) [inline lib-injection/requirements.json:9]: 2.28 is quite high for x86, do you know what symbol/dependency requires it?

running `nm -D library_file | grep U` will give you that answer

Comment (lloeki) [reply lib-injection/requirements.json:9]: The pipeline is using `registry.ddbuild.io/images/mirror/buildpack-deps:buster` as a base: https://github.com/DataDog/dd-trace-rb/blob/1ec390006b42b54a930776096f846c4f20651031/.gitlab/Dockerfile-3.2.2#L4

With which we get a glibc 2.28:

```
$ ldd --version
ldd (Debian GLIBC 2.28-10+deb10u2) 2.28
```

So [gems with native extensions are built against that](https://github.com/DataDog/dd-trace-rb/blob/1ec390006b42b54a930776096f846c4f20651031/.gitlab-ci.yml#L91).

The version requirement comes from there.

This results in the following max version referenced (== min version required):

```
tmp/arm64/3.2.0/extensions/aarch64-linux/3.2.0-static/datadog-2.4.0/datadog_profiling_loader.3.2.2_aarch64-linux.so: GLIBC_2.17
tmp/arm64/3.2.0/extensions/aarch64-linux/3.2.0-static/datadog-2.4.0/libdatadog_api.3.2_aarch64-linux.so: GLIBC_2.17
tmp/arm64/3.2.0/extensions/aarch64-linux/3.2.0-static/ffi-1.16.3/ffi_c.so: GLIBC_2.27
tmp/arm64/3.2.0/extensions/aarch64-linux/3.2.0-static/msgpack-1.7.3/msgpack/msgpack.so: GLIBC_2.17
tmp/arm64/3.2.0/gems/datadog-2.4.0/lib/datadog_profiling_loader.3.2.2_aarch64-linux.so: GLIBC_2.17
tmp/arm64/3.2.0/gems/datadog-2.4.0/lib/libdatadog_api.3.2_aarch64-linux.so: GLIBC_2.17
tmp/arm64/3.2.0/gems/ffi-1.16.3/lib/ffi_c.so: GLIBC_2.27
tmp/arm64/3.2.0/gems/libdatadog-13.1.0.1.0-aarch64-linux/vendor/libdatadog-13.1.0/aarch64-linux-musl/libdatadog-aarch64-alpine-linux-musl/lib/libdatadog_profiling.so: GLIBC_2.0
tmp/arm64/3.2.0/gems/libdatadog-13.1.0.1.0-aarch64-linux/vendor/libdatadog-13.1.0/aarch64-linux/libdatadog-aarch64-unknown-linux-gnu/lib/libdatadog_profiling.so: GLIBC_2.17
tmp/arm64/3.2.0/gems/libddwaf-1.14.0.0.0-aarch64-linux/vendor/libddwaf/libddwaf-1.14.0-linux-aarch64/lib/libddwaf.so: 
tmp/arm64/3.2.0/gems/msgpack-1.7.3/lib/msgpack/msgpack.so: GLIBC_2.17
```

`ffi` is the one using `GLIBC_2.27`. So we can lower it to 2.27 but until we move to building against a lower glibc (which comes with its own set of issues) we're stuck with that. FYI using [manylinux](https://github.com/pypa/manylinux) is under consideration.

Comment (raphaelgavache) [reply lib-injection/requirements.json:9]: if it helps you can pick agent build images or the ones used by the injector:
- [centos7 for amd](https://github.com/DataDog/auto_inject/blob/main/.gitlab/Dockerfile.rpm-x64)
- [for arm](https://github.com/DataDog/auto_inject/blob/main/.gitlab/Dockerfile.rpm-arm64)

Given that you download the source codes, won't building ffi on those machines improve your coverage? Or are there specific recent symbols that are required

Comment (lloeki) [reply lib-injection/requirements.json]: I think [`memfd_create`](https://www.man7.org/linux/man-pages/man2/memfd_create.2.html) needs glibc 2.27. 

It looks like `libffi` has fallback mechanisms if it is absent (see [Memory Usage](https://github.com/libffi/libffi/blob/0859f8431242d5adff21420b9cab538d2af527b5/doc/libffi.texi)), but it is an important security feature.

**Review by raphaelgavache — APPROVED** [review]

**Comment by lloeki** [comment]
Tests passed on this minimal one.

Screenshot 2024-10-30 at 18 35 12

**Comment by lloeki** [comment]
- Rebased.
- Undrafted for review.
- Once tests pass we can merge this, with further improvements down the road, notably covering for [INPLAT-103](https://datadoghq.atlassian.net/browse/INPLAT-103)

**Pull request merged**